### PR TITLE
[4.0] Helper factory

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for pr #32291.

### Summary of Changes
All over joomla are helpers manually loaded. Better to load them through a factory from the booted extension.

### Testing Instructions
- Open the dashboard.
- Add the following code to the file administrator/modules/mod_quickicon/src/Helper/QuickIconHelper.php:  
```php
public static function getDataAjax()
{
	$data = ['test'];

	return $data;
}
```
- Open the url /administrator/index.php?option=com_ajax&module=quickIcon&method=getData&format=json

### Actual result BEFORE applying this Pull Request
- Icons are loaded.
- Ajax url shows `{"success":true,"message":null,"messages":null,"data":["test"]}`

### Expected result AFTER applying this Pull Request
- Icons are loaded.
- Ajax url shows `{"success":false,"message":"The file at mod_quickIcon\/helper.php does not exist.","messages":null,"data":null}`

### Documentation Changes Required
Should be documented on the same place where the new extension setup is documented.

### CODE REVIEW

**Comment by laoneo** [comment]
My eyes still hurt from editing ajax.php!

**Comment by thednp** [comment]
I've looked into changes and I think it looks solid.

**Comment by laoneo** [comment]
@SharkyKZ why downvote?

**Comment by thednp** [comment]
Since my #30816 was pushed for 4.1, all of a sudden my interest in this dropped.

> @thednp if we can bring #32633 somehow to fly, then the same pattern can be applied to templates.

I don't see that comin any time soon this year.

**Comment by laoneo** [comment]
I understand you...

**Comment by joomdonation** [comment]
It is working as described. However, there are few things:

- You do not validate and make sure the method exists before calling the method. For legacy module, we throws LogicException
```php
$results = new LogicException(Text::sprintf('COM_AJAX_METHOD_NOT_EXISTS', $method . 'Ajax'), 404);
```
- Look like you bypass the extension enabled check like in the legacy module:

```php
if ($results === null && $moduleId && $table->load($moduleId) && $table->enabled)
```

I won't comment about the architecture change here because I'm not good at it and still trying to get myself familiar with architecture changes in J4.

**Comment by laoneo** [comment]
@joomdonation please comment inline in the code. Easier to understand.
@wilsonge do we need here also the release blocker label as the previous one had it?

File: components/com_ajax/ajax.php [hunk 1]
```diff
@@ -58,42 +58,51 @@
...
 
-			$class .= 'Helper';
-		}
-		else
-		{
-			$class = 'Mod' . ucfirst($module) . 'Helper';
-		}
+	if ($moduleInstance instanceof \Joomla\CMS\Helper\HelperFactoryInterface && $helper = $moduleInstance->getHelper($class))
+	{
+		$results = $helper->{$method . 'Ajax'}();
```
Comment (joomdonation) [inline components/com_ajax/ajax.php]: Here, you call the method without check and make sure that the method exists before calling (like what we do with none namespace helper module https://github.com/joomla/joomla-cms/blob/4.0-dev/components/com_ajax/ajax.php#L99)

You also do not check and make sure the module is enabled, like what we are doing here https://github.com/joomla/joomla-cms/blob/4.0-dev/components/com_ajax/ajax.php#L64 for none namespace module

Comment (laoneo) [reply components/com_ajax/ajax.php]: Can you have another look, restored most of the ajax.php stuff and now the boot call is inside the checks.

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 50828d9c056e92ba6313c6a2247d355cd86d2e51<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32633">issues.joomla.org/tracker/joomla-cms/32633</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 50828d9c056e92ba6313c6a2247d355cd86d2e51<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32633">issues.joomla.org/tracker/joomla-cms/32633</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32633">issues.joomla.org/tracker/joomla-cms/32633</a>.</sub>
